dpu-cni/pkgs/cniserver/cniserver_test.go

```diff
@@ -0,0 +1,72 @@
+package cniserver_test
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	"github.com/openshift/dpu-operator/dpu-cni/pkgs/cni"
+	"github.com/openshift/dpu-operator/dpu-cni/pkgs/cniserver"
+	"github.com/openshift/dpu-operator/dpu-cni/pkgs/cnitypes"
+
+	current "github.com/containernetworking/cni/pkg/types/100"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	utiltesting "k8s.io/client-go/util/testing"
+	"k8s.io/klog/v2"
+)
+
+func processRequest(request *cnitypes.Request) error {
+	// FIXME: Do actual work here.
+	klog.Infof("DEBUG: %v", request)
+	return nil
+}
+
+var _ = g.Describe("Cniserver", func() {
+	var tmpDir string
+	var plugin *cni.Plugin
+
+	var err error
+	// Create a tmp directory in the test container
+	tmpDir, err = utiltesting.MkTmpdir("cniserver")
+	defer os.RemoveAll(tmpDir)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	serverSocketPath := filepath.Join(tmpDir, cnitypes.ServerSocketName)
+	cniServer, err := cniserver.NewCNIServer(processRequest)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	go utilwait.Forever(func() {
+		cniServer.Start(tmpDir, serverSocketPath)
+	}, 0)
+
+	plugin = &cni.Plugin{SocketPath: serverSocketPath}
+
+	g.Context("CNI Server APIs", func() {
+		g.When("Normal ADD request", func() {
+			cniVersion := "0.4.0"
+			cniConfig := "{\"cniVersion\": \"" + cniVersion + "\",\"name\": \"dpucni\",\"type\": \"dpucni\"}"
+			cmdArgs := &skel.CmdArgs{
+				ContainerID: "fakecontainerid",
+				Netns:       "fakenetns",
+				IfName:      "fakeeth0",
+				Args:        "",
+				Path:        "fakepath",
+				StdinData:   []byte(cniConfig),
+			}
+			expectedResult := &current.Result{
+				CNIVersion: cniVersion,
+			}
+			os.Setenv("CNI_COMMAND", "ADD")
+			os.Setenv("CNI_CONTAINERID", cmdArgs.ContainerID)
+			os.Setenv("CNI_NETNS", cmdArgs.Netns)
+			os.Setenv("CNI_IFNAME", cmdArgs.IfName)
+			os.Setenv("CNI_PATH", cmdArgs.Path)
+			g.It("should get a correct response from the post request", func() {
+				resp, ver, err := plugin.PostRequest(cmdArgs)
+				o.Expect(err).NotTo(o.HaveOccurred())
+				o.Expect(ver).To(o.Equal(cniVersion))
+				o.Expect(resp.Result).To(o.Equal(expectedResult))
```

Comment:
What do we gain by having this test?
What does a passing test tell us?
If some commit starts to cause this to fail, what does that tell us about the config?

As per my understanding, `postRequest` is a translation from cni args to http request. It's a function that reads stuff and writes it somewhere else.

Reply:
What we gain: APIs should correctly parse the arguments. This should evolve as we use more of the arguments. This tests that any changes to the socket code won't affect anything. As this test, test both server and CNI shim side.
What a passing/failing test tells us:  As we do development we should really be aware to add more tests here to make sure relevant arguments are correctly expected on both the server and CNI shim side. The hope is that if we extend the arguments that we use the tests will start failing and the developer needs to add more test cases to make sure arguments are correctly passes between server and CNI shim. This is where `expectedResult` matters.
Yes `postRequest` comes from the CNI shim code side to send the CNI arguments over http.